### ISSUE TITLE
ListBoxField loses its value if it doesn't relate to a model property

### ISSUE THREAD

**Issue by robingram** [silverstripe/silverstripe-framework#8859]
## Affected Version

4.3

## Description

We added a `ListBoxField` to a form for editing a model in the CMS. This particular field didn't map directly to a property of the DataObject, it was processed in `onBeforeWrite` and saved into a different field. However, in the `onBeforeWrite` method the value for the `ListBoxField` was `null`.

We have a `TextareaField` on the form that also doesn't map to a property but that works.

It appears that the `saveInto` method of `FormField` calls `setCastedField` on the DO which will dynamically create the property if it isn't a database field. However, the overridden `saveInto` on `MultiSelectSield` will only attempt to set the value if it exists as a property on the object: https://github.com/silverstripe/silverstripe-framework/blob/4/src/Forms/MultiSelectField.php#L140.

This means that the behaviour is inconsistent between field types. It has also changed since SS3.x.

## Steps to Reproduce

In `getCMSFields` of a DataObject that has a model admin create a `ListboxField` with a set of static values. Add some code in `onBeforeWrite` to read the value of the list box. Edit an instance of the model in the CMS and select a value in the list box.

Expected: When `onBeforeWrite` is called the DO has a property with the same name as the `ListBoxField` that is populated with the value that was selected.

Actual: The value of the property is `null`.

## Workaround?

I'm looking to see if creating a setter method for the property functions as a workaround.

**Comment by robingram** [comment]
Creating a getter _and_ setter for the value works but the string encoding isn't great for multi-select boxes. It's necessary to `json_decode` the value too.

**Comment by maxime-rainville** [comment]
Just had a look. I confirmed the reported behaviour. However, I'm not convinced this is a problem.

I actually think it's a sensible behaviour to not pollute the record data for fields that don't exist on the DataObject.

If the information is meant to be save indirectly on the data object, the getter/setter approach sounds like a good one. If this is very specific to the form, then maybe the extra fields should be process on the edit form action.

**Comment by robingram** [comment]
That sounds reasonable. It's still worth noting though that this behaviour is inconsistent and most other field types _will_ set a property on the DataObject.

**Comment by maxime-rainville** [comment]
Agreed. We should adjust all form fields not to set values for non-existent field on DataObject.
